sites/cnfpt/src/backend/base/cache.py

```diff
@@ -0,0 +1,184 @@
+"""
+    RedisCache with a fallback cache to prevent denial of service if Redis is down
+    Freely inspired by django-cache-fallback
+
+    Credits:
+    - https://github.com/Kub-AT/django-cache-fallback/
+"""
+
+import logging
+
+from django.conf import settings
+from django.core.cache import caches
+from django.core.cache.backends.base import BaseCache
+
+from cms.cache import CMS_PAGE_CACHE_VERSION_KEY
+from cms.utils.conf import get_cms_setting
+from django_redis.cache import RedisCache
+
+DJANGO_REDIS_LOGGER = getattr(settings, "DJANGO_REDIS_LOGGER", __name__)
+logger = logging.getLogger(DJANGO_REDIS_LOGGER)
+
+
+class RedisCacheWithFallback(BaseCache):
+    """
+    BaseCache object with a redis_cache used as main cache
+    and used the "fallback" aliased cachewhich takes over
+    in case redis_cache is down.
+    """
+
+    def __init__(self, server, params):
+        """
+        Instantiate the Redis Cache with server and params
+        and retrieve the cache with alias "fallback"
+        """
+        super().__init__(params)
+        self._redis_cache = RedisCache(server, params)
+        self._fallback_cache = caches["fallback"]
+
+    def _call_with_fallback(self, method, *args, **kwargs):
+        """
+        Try first to exec provided method through Redis cache instance,
+        in case of failure, logger reports the exception and
+        the fallback cache takes over.
+        """
+        try:
+            next_cache_state = self._call_redis_cache(method, args, kwargs)
+            self._invalidate_fallback_cache()
+            return next_cache_state
+        except Exception as e:
+            logger.warning("[DEGRADED CACHE MODE] - Switch to fallback cache")
+            logger.exception(e)
+            return self._call_fallback_cache(method, args, kwargs)
+
+    def _call_redis_cache(self, method, args, kwargs):
+        """
+        Exec the provided method through the redis cache instance
+        """
+        return getattr(self._redis_cache, method)(*args, **kwargs)
+
+    def _call_fallback_cache(self, method, args, kwargs):
+        """
+        Exec the provided method through the fallback cache instance
+        """
+        return getattr(self._fallback_cache, method)(*args, **kwargs)
+
+    def _invalidate_fallback_cache(self):
+        """
+        From fallback cache, invalidate cms page cache.
+        """
+        version = self._fallback_cache.get(CMS_PAGE_CACHE_VERSION_KEY)
+        if version:
+            self._fallback_cache.set(
+                CMS_PAGE_CACHE_VERSION_KEY,
+                version + 1,
+                get_cms_setting("CACHE_DURATIONS")["content"],
+            )
```

Comment:
I'm thinking : this is going to increment the version VERY high ^^ Why not delete the value completed since we are not trying to bump the cache version but to reset it completely?

Reply:
You are right! In this way, we make just one operation 🎉